src/main.rs

```diff
@@ -238,3 +256,125 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+                write!(f, "error setting update announcement: {err}")
+            }
+            SetOnionBitError::GetInfoErr(err) => write!(f, "get_info error: {err}"),
+        }
+    }
+}
+
+#[async_trait]
+trait NodeInfoRetriever {
+    async fn get_info(
+        &mut self,
+        request: tonic_lnd::lnrpc::GetInfoRequest,
+    ) -> Result<tonic_lnd::lnrpc::GetInfoResponse, SetOnionBitError>;
+}
+
+struct GetInfoClient<'a> {
+    client: &'a mut LightningClient,
+}
+
+#[async_trait]
+impl<'a> NodeInfoRetriever for GetInfoClient<'a> {
+    async fn get_info(
+        &mut self,
+        request: tonic_lnd::lnrpc::GetInfoRequest,
+    ) -> Result<tonic_lnd::lnrpc::GetInfoResponse, SetOnionBitError> {
+        match self.client.get_info(request).await {
+            Ok(resp) => Ok(resp.into_inner()),
+            Err(status) => Err(SetOnionBitError::GetInfoErr(status)),
+        }
+    }
+}
+
+/// UpdateNodeAnnouncement defines the peer functionality needed to set the onion
+/// messaging feature bit.
+#[async_trait]
+trait UpdateNodeAnnouncement {
+    async fn update_node_announcement(
+        &mut self,
+        request: tonic_lnd::peersrpc::NodeAnnouncementUpdateRequest,
+    ) -> Result<(), SetOnionBitError>;
+}
+
+struct UpdateNodeAnnClient<'a> {
+    peers_client: &'a mut PeersClient,
+}
+
+#[async_trait]
+impl<'a> UpdateNodeAnnouncement for UpdateNodeAnnClient<'a> {
+    async fn update_node_announcement(
+        &mut self,
+        request: tonic_lnd::peersrpc::NodeAnnouncementUpdateRequest,
+    ) -> Result<(), SetOnionBitError> {
+        match self.peers_client.update_node_announcement(request).await {
+            Ok(_) => Ok(()),
+            Err(status) => {
+                if status.code() == Code::Unimplemented {
+                    return Err(SetOnionBitError::UnimplementedPeersService);
+                }
+
+                return Err(SetOnionBitError::UpdateAnnouncementErr(status));
+            }
+        }
+    }
+}
+
+/// Sets the onion messaging feature bit (described in this PR:
+/// https://github.com/lightning/bolts/pull/759/), to signal that we support
+/// onion messaging. This needs to be done every time LND starts up, because LND
+/// does not currently persist the custom feature bits that are set via the RPC.
+async fn set_feature_bit(
+    client: &mut impl NodeInfoRetriever,
+    peers_client: &mut impl UpdateNodeAnnouncement,
+) -> Result<(), SetOnionBitError> {
+    let update = tonic_lnd::peersrpc::UpdateFeatureAction {
+        action: i32::from(tonic_lnd::peersrpc::UpdateAction::Add),
+        feature_bit: ONION_MESSAGES_OPTIONAL as i32,
+    };
+    let feature_updates = vec![update];
+
+    peers_client
+        .update_node_announcement(tonic_lnd::peersrpc::NodeAnnouncementUpdateRequest {
+            feature_updates,
+            color: Default::default(),
+            alias: Default::default(),
+            address_updates: Default::default(),
```

Comment:
Can be flattened into `..Default::default()` to auto-fill the remaining fields with defaults.